chromium_src/third_party/blink/renderer/modules/storage/dom_window_storage.cc

```diff
@@ -31,25 +42,209 @@ void MaybeClearAccessDeniedException(StorageArea* storage,
+    ProvideTo(window, supplement);
+  }
+  return *supplement;
+}
+
+// static
+StorageArea* BraveDOMWindowStorage::sessionStorage(
+    LocalDOMWindow& window,
+    ExceptionState& exception_state) {
+  return From(window).sessionStorage(exception_state);
 }
 
 // static
-StorageArea* BraveDOMWindowStorage::localStorage(LocalDOMWindow& window,
-                                            ExceptionState& exception_state) {
+StorageArea* BraveDOMWindowStorage::localStorage(
+    LocalDOMWindow& window,
+    ExceptionState& exception_state) {
+  return From(window).localStorage(exception_state);
+}
+
+BraveDOMWindowStorage::BraveDOMWindowStorage(LocalDOMWindow& window)
+    : Supplement<LocalDOMWindow>(window) {}
+
+StorageArea* BraveDOMWindowStorage::sessionStorage(
+    ExceptionState& exception_state) {
+  LocalDOMWindow* window = GetSupplementable();
   auto* storage =
-      DOMWindowStorage::From(window).localStorage(exception_state);
-  MaybeClearAccessDeniedException(storage, window, &exception_state);
-  return storage;
+      DOMWindowStorage::From(*window).sessionStorage(exception_state);
+
+  MaybeClearAccessDeniedException(storage, *window, &exception_state);
+  return ephemeralSessionStorage(storage);
+}
+
+StorageArea* BraveDOMWindowStorage::ephemeralSessionStorage(
+    StorageArea* non_ephemeral_storage) {
+  if (!base::FeatureList::IsEnabled(features::kBraveEphemeralStorage))
+    return non_ephemeral_storage;
+
+  LocalDOMWindow* window = GetSupplementable();
+  if (!window->IsCrossSiteSubframe())
+    return non_ephemeral_storage;
+
+  // If we were not able to create non-ephemeral session storage for this
+  // window, then don't attempt to create an ephemeral version.
+  if (!non_ephemeral_storage)
+    return nullptr;
+
+  if (ephemeral_session_storage_)
+    return ephemeral_session_storage_;
+
+  Page* page = window->GetFrame()->GetDocument()->GetPage();
+  EphemeralStorageNamespaces* namespaces =
+      EphemeralStorageNamespaces::From(page, window);
+  if (!namespaces)
+    return nullptr;
+
+  auto storage_area =
+      namespaces->session_storage()->GetCachedArea(window->GetSecurityOrigin());
+
+  ephemeral_session_storage_ =
+      StorageArea::Create(window->GetFrame(), std::move(storage_area),
+                          StorageArea::StorageType::kSessionStorage);
+  return ephemeral_session_storage_;
+}
+
+StorageArea* BraveDOMWindowStorage::localStorage(
+    ExceptionState& exception_state) {
+  LocalDOMWindow* window = GetSupplementable();
+  auto* storage = DOMWindowStorage::From(*window).localStorage(exception_state);
+
+  MaybeClearAccessDeniedException(storage, *window, &exception_state);
+  return ephemeralLocalStorage(storage);
+}
+
+StorageArea* BraveDOMWindowStorage::ephemeralLocalStorage(
+    StorageArea* non_ephemeral_storage) {
+  if (!base::FeatureList::IsEnabled(features::kBraveEphemeralStorage))
+    return non_ephemeral_storage;
```

Comment:
I'm thinking that maybe we can move these checks to the upper level? Or somehow change the function name, because now it is actually not `ephemeralLocalStorage`, but "ephemeral local storage if things are good, otherwise normal local storage"

Reply:
:+1: